feat: add theme variant service

d-loose opened the pull request:
Adds a `ThemeVariantService` to ubuntu-provision that loads some configuration options from the `ConfigService` and creates a `ThemeVariant`, which contains customized yaru themes and a window title.
Example configuration:
```yaml
bootstrap:
  accent-color: '#0000ff'
  elevated-button-color: '#ff0000'
  elevated-button-text-color: '#00ff00'
  window-title: 'My custom window title'
```
(the color parameters are already supported by yaru.dart without any further customization needed).

I'm sure we'll want to adjust details like config file location, scope, etc later on.

Ref #205
UDENG-1762

spydon commented on packages/ubuntu_provision/lib/src/services/theme_variant_service.dart:
```diff
@@ -0,0 +1,77 @@
+
+import '../../services.dart';
+
+/// @internal
```
```suggestion
@internal
```
`meta` will have to be imported too.

Will this be used outside of this file? Otherwise it could just be made private.

d-loose replied:
I just wanted to be consistent with all the other files that follow the exact same pattern. Is it worth changing this across the entire project?

spydon replied:
What does a dartdoc with @internal in it even mean? Surely that doesn't affect the analyzer? :thinking:

d-loose replied:
I have no idea, it's probably unnecessary and doesn't do anything, since the `log` symbols need to be explicitly hidden from the barrel exports

spydon replied:
Let's make them private where possible and `@internal` at the other places then, I can do that in a separate PR. :)

tim-hm commented:
Great start @d-loose 🙏 🤩  I think we should start with a simpler approach and evolve if/when required; after all YAGNI and its harder to retract the feature once its live! I think we should require/assume that if a `whitelabel.yaml` was provided then it will be available for every stage -- ie we copy it from bootstrap to the target so its automatically picked up during 4/5/6.

A few suggestions on the schema:

1. The styling override could be against a global theme key and independent of the stage we're on:

```
theme:
  accent-color: '#0000ff'
  elevated-button-color: '#ff0000'
  elevated-button-text-color: '#00ff00'
```

2. For overriding the window title and image we can then work by stage / screen:

```
screens:
   name-of-step:
     image: <path-to-image>
     title: <custom title>
```

3. Finally, for turning on/off individual screens we simply provide false aginst the step id/name:

```
screens:
   name-of-step: false
```

WDYT?

d-loose commented:
> I think we should require/assume that if a `whitelabel.yaml` was provided then it will be available for every stage -- ie we copy it from bootstrap to the target so its automatically picked up during 4/5/6.

Yes, that makes sense. Will we need more than one configuration file for the UI in the end? We're currently assuming there's a single `ubuntu-provision.yaml`, which we could just rename to `whitelabel.yaml` if that's all we need. All the other configuration files (autoinstall, dry-run, etc) are read by subiquity.


> A few suggestions on the schema:
> ...
> WDYT?

Looks good!

tim-hm commented:
> Will we need more than one configuration file for the UI in the end? 

I think we should be working towards a single whitelabel.yaml file for the moment.

d-loose commented:
We're currently using `bootstrap` and `init` scopes in the config file - I'd restructure this and adopt the suggested schema in a separate PR.

d-loose commented:
Since I did some refactoring, it seems to make more sense to include the restructuring in this PR.
On the surface, not much has changed:
```yaml
theme:
  accent-color: '#0000ff'
  elevated-button-color: '#ff0000'
  elevated-button-text-color: '#00ff00'
  window-title: 'My custom window title'
```
can be used to configure the theme for both ubuntu-bootstrap and ubuntu-init now (I've removed the separate config scopes).

I've added `freezed` and `json_serialize` to ubuntu_provision for a more structured config parsing approach. It might make sense to move the `ThemeConfig` class out of the service at a later point.
